draft-ietf-mls-protocol.md

```diff
@@ -2833,7 +2833,7 @@ a Commit message before sending application data. This ensures, for example,
 that any members whose removal was proposed during the epoch are actually
 removed before any application data is transmitted.
 
-The sender of a Commit MUST include all valid proposals that it has received
+The sender of a Commit SHOULD include all valid proposals that it has received
```

Comment:
I don't really agree with downgrading this from a MUST to a SHOULD. Quoting RFC 2119:

```
Imperatives of the type defined in this memo must be used with care
and sparingly.  In particular, they MUST only be used where it is
actually required for interoperation or to limit behavior which has
potential for causing harm (e.g., limiting retransmisssions)
```

This is a case where we're limiting potentially harmful behavior (intentionally dropping proposals and forcing others to retransmit) such that implementations that don't follow the MUST are non-compliant with the spec.

Reply:
I see your point, my primary concern was reducing how often commits are rejected due to the async nature of the protocol since those packets likely have a higher computational and bandwidth cost.

"A group member that has observed one or more proposals within an epoch MUST send a Commit message" also plays into this choice. If a group is receiving an uptick in proposals, which tends to happen when they are first created and a large number of users in size, then commits will fail and block members from sending messages until a new commit is generated and uploaded. When you are using the protocol on something like a mobile app on a bad connection this becomes a problem due to high amounts of latency between message send and receive.

Reply:
Commits wouldn't be rejected if they don't have 100% of the proposals. There are valid reasons a Commit would not have all proposals: lossy delivery channel, race condition. So like in the case where 1.) I start preparing a Commit, 2.) You send a new proposal, 3.) I send my Commit without your proposal -- your proposal would be dropped and you'd have the option to re-send it in the new epoch. There's not a technical enforcement of this "MUST", its more about policy.

I believe a Commit would only be rejected if it was sent at the same time as another Commit, and the Delivery Service chose to use the other Commit instead. That would be less common because it would only happen when two users are trying to send application data at the same time.

Reply:
This is why I think SHOULD is appropriate here -- I hesitate to have a MUST-level requirement for behavior that is not observable to remote parties.  After all, if it's not observable, then it can hardly be a requirement for interop!

Reply:
In my experience it is totally reasonable to split behavior into sending/receiving roles for example. It's fine to say that when you send a commit you MUST include all valid proposals you have received, as long as the receiving behavior makes it clear what to do (and under what circumstances the receiver might see a proposal not in the commit: ex: invalid, or hadn't arrived yet).

If we make this a SHOULD, we need an extra sentence explaining under what circumstances the sender might choose not to include a specific proposal. SHOULD's are always more complicated. My general IETF advice is to use them sparingly.

Reply:
I can live with that.  Then maybe the way to resolve the confusion @tomleavy notes is to add a note here that receivers SHOULD NOT enforce that everything they sent is included?